src/stripeWorkspaceState.ts

```diff
@@ -89,9 +91,48 @@ export function addLogDetails(
   extensionContext.workspaceState.update(logDetailsKey, logDetailsMap);
 }
 
-export function retrieveLogDetails(extensionContext: vscode.ExtensionContext, logId: string) {
+export async function retrieveLogDetails(extensionContext: vscode.ExtensionContext, logId: string, daemonClient: StripeCLIClient) {
   const logDetailsMap = getLogDetailsMap(extensionContext);
-  return logDetailsMap.get(logId);
+  const logDetails = logDetailsMap.get(logId);
+
+  // if insight has not been retrieved or previously failed to be retrieved, then retrieve it
+  if (!('insight' in logDetails) || logDetails.insight.includes('Failed to retrieve insight') || logDetails.insight.includes('Please check back later')) {
+    const insight = await getIntegrationInsight(logId, daemonClient);
+    logDetails.insight = insight;
+    addLogDetails(extensionContext, logId, logDetails);
+  }
+
+  return logDetails;
+}
+
+async function getIntegrationInsight(logId: string, daemonClient: StripeCLIClient): Promise<string> {
+  const request = new IntegrationInsightRequest();
+  request.setLog(logId);
+
+  const integrationInsight = await new Promise<string>((resolve, reject) => {
+    daemonClient.integrationInsight(request, (error: any, response: any) => {
+      if (error) {
+        if (error.code === 12) {
+          // https://grpc.github.io/grpc/core/md_doc_statuscodes.html
+          // 12: UNIMPLEMENTED
```

Comment:
Nit: the grpc library should have these error codes built in as constants.

Reply:
i will use the grpc constants